File: libs/ui-app-state-manager/src/main/kotlin/mono/ui/appstate/AppThemeManager.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.ui.appstate

import mono.lifecycle.LifecycleOwner
import mono.store.manager.StorageDocument
import mono.store.manager.StoreKeys
import mono.ui.theme.ThemeManager
import mono.ui.theme.ThemeMode
import org.w3c.dom.Element

/**
 * A class for managing theme
 */
internal class AppThemeManager(
    private val themeManager: ThemeManager,
    private val settingsDocument: StorageDocument = StorageDocument.get(StoreKeys.SETTINGS)
) {

    init {
        val themeMode = settingsDocument.get(StoreKeys.THEME_MODE)
            ?.let(ThemeMode::valueOf)
            ?: ThemeMode.DARK
        val themeManager = ThemeManager.getInstance()
        themeManager.setTheme(themeMode)
    }

    fun observeTheme(
        appLifecycleOwner: LifecycleOwner,
        documentElement: Element,
        forceUiUpdate: () -> Unit
    ) {
        themeManager.themeModeLiveData.observe(appLifecycleOwner) {
            documentElement.className = when (it) {
                ThemeMode.LIGHT -> THEME_LIGHT
                ThemeMode.DARK -> THEME_DARK
            }
            forceUiUpdate()
            settingsDocument.set(StoreKeys.THEME_MODE, it.name)
        }

        settingsDocument.setObserver(StoreKeys.THEME_MODE) { _, _, newValue ->
            val themeMode = newValue?.let(ThemeMode::valueOf) ?: ThemeMode.DARK
            themeManager.setTheme(themeMode)
        }
    }

    companion object {
        private const val THEME_LIGHT = "light"
        private const val THEME_DARK = "dark"
    }
}

File: monosketch-svelte/src/lib/mono/ui-state-manager/app-theme-manager.ts
import { ThemeManager, ThemeMode } from '$mono/theme';
import { StorageDocument, StoreKeys } from '$mono/store-manager';
import { Flow, LifecycleOwner } from '$mono/flow';

/**
 * A class for managing theme of the app, including observing theme change from
 * the settings.
 */
export class AppThemeManager {
    private settingsDocument: StorageDocument = StorageDocument.get(StoreKeys.SETTINGS);
    private themeManager: ThemeManager = ThemeManager.getInstance();

    themeModeFlow: Flow<ThemeMode> = this.themeManager.themeModeFlow;

    constructor() {
        const themeMode = this.settingsDocument.get(StoreKeys.THEME_MODE);
        if (themeMode) {
            this.themeManager.setTheme(themeMode as ThemeMode);
        }
    }

    setTheme = (themeMode: ThemeMode): void => {
        this.themeManager.setTheme(themeMode);
    };

    observeTheme = (appLifecycleOwner: LifecycleOwner, onThemeChange: () => void): void => {
        this.themeManager.themeModeFlow.observe(appLifecycleOwner, (themeMode) => {
            document.documentElement.className = themeMode;
            onThemeChange();

            this.settingsDocument.set(StoreKeys.THEME_MODE, themeMode);
        });

        this.settingsDocument.setObserver(StoreKeys.THEME_MODE, {
            onChange: (key, oldValue, newValue) => {
                if (newValue) {
                    this.themeManager.setTheme(newValue as ThemeMode);
                }
            },
        });
    };
}
